### PR TITLE
Support SkyViper V2450

### PR DESCRIPTION
**Pull request by tridge**
This adds full support for the SkyViper 2450. It incorporates driver changes (some of which have separate PRs), plus additions to ArduCopter, new temperature calibration code and new declination code.
To build this firmware use:
./waf configure --board skyviper-v2450
it builds under ChibiOS

### CODE REVIEW

**Comment by rmackay9** [comment]
Looks like it's failing travis because "degrees C" appears in the parameter descriptions..

File: libraries/AP_Compass/Compass_PerMotor.cpp [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    // @Param: 4_Y
+    // @DisplayName: Compass per-motor4 Y
+    // @Description: Compensation for Y axis of motor4
+    // @User: Advanced
+
+    // @Param: 4_Z
+    // @DisplayName: Compass per-motor4 Z
+    // @Description: Compensation for Z axis of motor4
+    // @User: Advanced
+    AP_GROUPINFO("4",  6, Compass_PerMotor, compensation[3], 0),
```
Comment (davidbuzz) [inline libraries/AP_Compass/Compass_PerMotor.cpp]: compass compensation per-motor is limited to 4 motors?

File: libraries/AP_Compass/Compass_PerMotor.cpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+            SRV_Channels::find_channel(SRV_Channel::k_motor4, motor_map[3])) {
+            have_motor_map = true;
+        }
+    }
+    if (!have_motor_map) {
+        return 0;
+    }
+    
+    // this currently assumes first 4 channels. 
+    uint16_t pwm = hal.rcout->read_last_sent(motor_map[motor]);
```
Comment (davidbuzz) [inline libraries/AP_Compass/Compass_PerMotor.cpp]: assume is a lot like "TODO", is this safe-in-practice?

File: libraries/AP_Compass/Compass_PerMotor.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  per-motor compass compensation
+ */
+
+#include <AP_Math/AP_Math.h>
+
+
+class Compass;
+
+// per-motor compass compensation class. Currently tied to quadcopters
+// only, and single magnetometer
```
Comment (davidbuzz) [inline libraries/AP_Compass/Compass_PerMotor.h]: maybe raise this known limitation as an issue for future revisit ?

File: libraries/AP_Radio/AP_Radio.cpp [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+uint32_t AP_Radio::get_tx_version(void)
+{
+    if (driver) {
+        return driver->get_tx_version();
+    }
+    return 0;
+}
+
+// set the 2.4GHz wifi channel used by companion computer, so it can be avoided
+void AP_Radio::set_wifi_channel(uint8_t channel)
```
Comment (davidbuzz) [inline libraries/AP_Radio/AP_Radio.cpp]: this is more of a "avoid_wifi_channel" or "set_wifi_channel_to_avoid"  than "set_wifi_channel", causual observer/s might think this function *uses* the wifi channel.

File: libraries/AP_Radio/AP_Radio_backend.h [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    virtual bool reset(void) = 0;
+    
+    // send a packet
+    virtual bool send(const uint8_t *pkt, uint16_t len) = 0;
+
+    // start bind process as a receiver
+    virtual void start_recv_bind(void) = 0;
+
+    // return time in microseconds of last received R/C packet
+    virtual uint32_t last_recv_us(void) = 0;
```
Comment (davidbuzz) [inline libraries/AP_Radio/AP_Radio_backend.h]: A uint32 counting microseconds will be able to count up to about 35 minutes. It's common for people to power up some of their equipment but not all of it (such as while waiting for gps lock, or making coffee).  Is this going to potentially wrap-around if the TX is left turned off for more than ~35minutes.. ?

File: libraries/AP_Radio/AP_Radio_cc2500.cpp [hunk 6]
```diff
@@ -0,0 +1,874 @@
...
+    }
+
+    ch = get_tx_pps_chan();
+    if (ch > 0) {
+        dsm.pwm_channels[ch-1] = dsm.tx_pps;
+        dsm.num_channels = MAX(dsm.num_channels, ch);
+    }
+#endif
+    
+    if (now - last_pps_ms > 1000) {
```
Comment (davidbuzz) [inline libraries/AP_Radio/AP_Radio_cc2500.cpp]: hardcoded threshold of 1000 , maybe a #define, or a parameter?

File: libraries/AP_Radio/AP_Radio_cc2500.cpp [hunk 7]
```diff
@@ -0,0 +1,874 @@
...
+    protocolState = STATE_BIND_TUNING;
+    chan_count = 0;
+    packet_timer = AP_HAL::micros();
+    chEvtSignal(_irq_handler_ctx, EVT_BIND);
+    Debug(1,"Starting bind\n");
+}
+
+// handle a data96 mavlink packet for fw upload
+void AP_Radio_cc2500::handle_data_packet(mavlink_channel_t chan, const mavlink_data96_t &m)
+{
```
Comment (davidbuzz) [inline libraries/AP_Radio/AP_Radio_cc2500.cpp]: not impl, clearly.

File: libraries/AP_Radio/AP_Radio_cc2500.cpp [hunk 8]
```diff
@@ -0,0 +1,874 @@
...
+
+    if (!bind_storage.read_block(&info, 0, sizeof(info)) || info.magic != bind_magic) {
+        return false;
+    }
+    
+    bindTxId[0] = info.bindTxId[0];
+    bindTxId[1] = info.bindTxId[1];
+    bindOffset = info.bindOffset;
+    listLength = info.listLength;
+    memcpy(bindHopData, info.bindHopData, sizeof(bindHopData));
```
Comment (davidbuzz) [inline libraries/AP_Radio/AP_Radio_cc2500.cpp]: After blindly memcpy()-ing a chunk of data from ( possiby unreliable) storage into this struct, is it worth defensively checking to make sure that the data is 'good' before we trust it?

File: libraries/AP_Radio/AP_Radio_cypress.cpp [hunk 9]
```diff
@@ -0,0 +1,1748 @@
...
+#if CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
+    if(_irq_handler_ctx != nullptr) {
+        AP_HAL::panic("AP_Radio_cypress: double instantiation of irq_handler\n");
+    }
+    chVTObjectInit(&timeout_vt);
+    _irq_handler_ctx = chThdCreateStatic(_irq_handler_wa,
+                     sizeof(_irq_handler_wa),
+                     TIMEOUT_PRIORITY,        /* Initial priority.    */
+                     irq_handler_thd,  /* Thread function.     */
+                     NULL);                     /* Thread parameter.    */
```
Comment (davidbuzz) [inline libraries/AP_Radio/AP_Radio_cypress.cpp]: I thought the point of the 'hal' was to hid this sort of OS specific thing from the ArduPilot code.   Maybe this chibios "virtual timer" chVTObjectInit  and 'Thread'  chThdCreateStatic  activities should be put in a hal->timer() or hal->thread() or something less platform specific?

File: libraries/AP_Radio/AP_Radio_cypress.cpp [hunk 10]
```diff
@@ -0,0 +1,1748 @@
...
+        dsm.num_channels = MAX(dsm.num_channels, ch);
+    }
+
+    ch = get_tx_pps_chan();
+    if (ch > 0) {
+        dsm.pwm_channels[ch-1] = dsm.tx_pps;
+        dsm.num_channels = MAX(dsm.num_channels, ch);
+    }
+    
+    if (now - last_debug_print_ms > 1000) {
```
Comment (davidbuzz) [inline libraries/AP_Radio/AP_Radio_cypress.cpp]: another hardcoded 1000 value, see earlier comment on a #define or parameter?

File: libraries/AP_Radio/AP_Radio_cypress.cpp [hunk 11]
```diff
@@ -0,0 +1,1748 @@
...
+    Debug(1, "Cypress: radio_init done\n");
+
+    start_receive();
+
+    // setup handler for rising edge of IRQ pin
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4
+    stm32_gpiosetevent(CYRF_IRQ_INPUT, true, false, false, irq_radio_trampoline);
+#elif CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
+    hal.gpio->attach_interrupt(HAL_GPIO_RADIO_IRQ, trigger_irq_radio_event, HAL_GPIO_INTERRUPT_RISING);
+#endif
```
Comment (davidbuzz) [inline libraries/AP_Radio/AP_Radio_cypress.cpp]: this is another one of those cases where a hal should mean this block of #if code wouldn't be necerssary. 
Can't hal->attach_interrupt(....) be made smart enough to know if it's on px4, and then cal stm32_gpiosetevent behin the scenes?

File: libraries/AP_Radio/AP_Radio_cypress.cpp [hunk 12]
```diff
@@ -0,0 +1,1748 @@
...
+    }
+#else
+    write_register(CYRF_XACT_CFG, CYRF_MODE_SYNTH_TX | CYRF_FRC_END);
+    write_register(CYRF_RX_ABORT, 0);
+    transmit16(pkt);
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4
+    hrt_call_after(&wait_call, 10000, (hrt_callout)irq_timeout_trampoline, nullptr);
+#elif CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
+    //Note: please review Frequency in chconf.h to ensure the range of wait
+    chVTSet(&timeout_vt, MS2ST(10), trigger_timeout_event, nullptr);
```
Comment (davidbuzz) [inline libraries/AP_Radio/AP_Radio_cypress.cpp]: can these #ifdefs be moved into the hal-> ?

**Comment by davidbuzz** [comment]
I have reviewd, and put in a bunch of inline comments. but it's a +1 from me here for merge.

File: libraries/AP_Compass/Compass_PerMotor.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_Compass/Compass_PerMotor.cpp]: yes, for now, happy to extend it if there is demand

File: libraries/AP_Compass/Compass_PerMotor.h (hunk 3 above)
Comment (tridge) [reply libraries/AP_Compass/Compass_PerMotor.h]: worth discussing on a dev call

File: libraries/AP_Radio/AP_Radio_cypress.cpp (hunk 9 above)
Comment (tridge) [reply libraries/AP_Radio/AP_Radio_cypress.cpp]: yes indeed - and Sid has some patches to add APIs for this

**Comment by WickedShell** [comment]
@tridge It looks like this include the per-motor compass stuff which is already a it's own PR #6448 and has open review comments that don't seem to be addressed in this PR either. Just tagging it that there is some review that was done from a different location on that part of this.

**Comment by rmackay9** [comment]
I know it's slightly bad behaviour but I plucked out the flight mode channel change into master.  I did this because a user was pinging me about it so it gave me the opportunity to say, "it's done!".
https://github.com/ArduPilot/ardupilot/commit/89674482a717bd280c594c408c6feec7d569fac9

Sadly, I slightly modified the commit to remove the addition of the #define for TOY_MODE.  Sorry, I probably should have left that part in to avoid the merge conflict but I didn't for some reason.

**Comment by amilcarlucas** [comment]
Can the SkyViper be built under both NuttX and ChibiOS ?
Or is it ChibiOS only ?

**Comment by rmackay9** [comment]
@amilcarlucas, my guess is it's both but I'm not sure. @tridge and @bugobliterator know for sure

**Comment by bugobliterator** [comment]
@amilcarlucas  it should support both Nuttx and ChibiOS

**Comment by tridge** [comment]
merged after discussion with Randy
many thanks for input today from Peter, Randy and Michael

**Pull request merged**